server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -558,12 +558,25 @@ public Mapping parseMapping(String mappingType, Map<String, Object> mappingSourc
         }
     }
 
-    public static Mapping mergeMappings(DocumentMapper currentMapper, Mapping incomingMapping, MergeReason reason) {
+    public Mapping mergeMappings(Mapping incomingMapping, MergeReason reason) {
+        return mergeMappings(this.mapper, incomingMapping, reason, Long.MAX_VALUE);
+    }
+
+    static Mapping mergeMappings(
+        DocumentMapper currentMapper,
+        Mapping incomingMapping,
+        MergeReason reason,
+        long maxFieldsToAddDuringMerge
+    ) {
         Mapping newMapping;
         if (currentMapper == null) {
-            newMapping = incomingMapping;
+            if (MappingLookup.fromMapping(incomingMapping).exceedsLimit(maxFieldsToAddDuringMerge, 0)) {
+                newMapping = Mapping.EMPTY.merge(incomingMapping, reason, maxFieldsToAddDuringMerge);
+            } else {
+                newMapping = incomingMapping;
```

Comment:
Yes, we could do that. The else-branch is basically an optimization that prevents a merge process if it isn't necessary. But I'm equally happy to simplify the code.

Reply:
ok, I was maybe making the wrong assumption that that merge would mostly be a no-op.

Reply:
Ok, refreshing my memory: the current mapper is null when the index has no mappings, and that conditional aims to cover the scenario where the first document getting indexed goes already beyond the limit or not. I think I would remove this optimization and always call EMPTY.merge. Perhaps we could think about overriding `merge` for that specialized empty mapping and not doing any work, but it's final and I am not sure we want to make it non-final.

By the way, how are we going to make sure that this is not exercised by static mappings as well? It currently is, isn't it? That reminds me we have so many different merge methods and that makes things very difficult to follow. Which merge is used where? I will need to review that bit once we get there.

Reply:
> I think I would remove this optimization and always call EMPTY.merge.

I've implemented that and it actually revealed a bug that was caught by a lot of test cases. Good call to always call the same code path so that there's not a code path that's rarely executed. I've fixed the bug in 48e2f33.

> Perhaps we could think about overriding merge for that specialized empty mapping and not doing any work, but it's final and I am not sure we want to make it non-final.

The merge is short-circuited in `MergeResult#addNewObjectMapper` so it won't recurse all the way down. Seems like that's enough of an optimization.

Reply:
Tell me more about the bug? I was going to ask you why you went from Mapping.merge to root.withoutMapper.merge. It's subtle!

Reply:
If `enabled` is set to `false` in the root object mapper, the following exception was thrown:

https://github.com/elastic/elasticsearch/blob/9e074b8e7521129504024cc2b7b446c3485611da/server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java#L499

That's because `Mapper.EMPTY` has an implicit value for `enabled` of `true`. Thus, you can't merge `Mapper.EMPTY` with mappings that have set `enabled` to `false`.